operator/internal/decommissioning/statefulset_decomissioner.go

```diff
@@ -0,0 +1,753 @@
+				brokersToIgnore = append(brokersToIgnore, downedNode)
+				continue
+			}
+
+			if strings.Contains(err.Error(), "does not exist") {
+				// delete the node from our sets
+				downedNodes.Delete(downedNode)
+				continue
+			}
+
+			log.Error(err, "fetching decommission status")
+			return false, err
+		}
+
+		if status.Finished {
+			// skip since we have already decommissioned it, so it should no longer
+			// show up in the health overview
+			continue
+		}
+
+		// add the brokers to the list of what needs to be decommissioned
+		currentlyDecommissioningBrokers = append(currentlyDecommissioningBrokers, downedNode)
+	}
+
+	sortBrokers := func(set []int) []int {
+		// sort by simple node id
+		sort.SliceStable(set, func(i, j int) bool {
+			return set[i] < set[j]
+		})
+		return set
+	}
+
+	// NB: here is where we need to start to check the cache for any
+	// nodes that meet the threshold, otherwise they should just be considered
+	// ignored as well
+
+	// mark any brokers that currently meet our criteria for decommissioning
+	s.delayedBrokerIDCache.Filter(setCacheKey, brokersToDecommission...)
+	for _, broker := range brokersToDecommission {
+		s.delayedBrokerIDCache.Mark(setCacheKey, broker)
+	}
+
+	healthyBrokers := sortBrokers(healthyNodes.Values())
+	brokersToDecommission = sortBrokers(brokersToDecommission)
+	brokersToIgnore = sortBrokers(brokersToIgnore)
+	currentlyDecommissioningBrokers = sortBrokers(currentlyDecommissioningBrokers)
+
+	formatBrokerList := func(set []int) string {
+		return strings.Join(functional.MapFn(strconv.Itoa, set), ", ")
+	}
+
+	log.V(traceLevel).Info(fmt.Sprintf(
+		"healthy brokers: [%s], ignored: [%s], to decommission: [%s], decommissioning: [%s]",
+		formatBrokerList(healthyBrokers),
+		formatBrokerList(brokersToIgnore),
+		formatBrokerList(brokersToDecommission),
+		formatBrokerList(currentlyDecommissioningBrokers),
+	))
+
+	// we marked the brokers to decommission above, but don't attempt a decommission yet
+	// because something is already in progress
+
+	if len(currentlyDecommissioningBrokers) != 0 {
+		// we skip decommissioning our next broker since we already have some node decommissioning in progress
+		return true, nil
+	}
+
+	// now attempt to decommission something, if we have actually done something, then requeue
+	for _, broker := range brokersToDecommission {
+		decommissioned, err := s.delayedBrokerIDCache.Process(setCacheKey, broker, func() error {
+			// only record the event here since this is when we trigger a decommission
+			s.recorder.Eventf(set, corev1.EventTypeNormal, eventReasonBroker, "brokers needing decommissioning: [%s], decommissioning: %d", formatBrokerList(brokersToDecommission), brokersToDecommission[0])
+
+			if err := adminClient.DecommissionBroker(ctx, brokersToDecommission[0]); err != nil {
+				log.Error(err, "decommissioning broker", "broker", brokersToDecommission[0])
+				return err
+			}
+
+			return nil
+		})
```

Comment:
In this `for` loop we should process `broker`, but the `Event` and `DecommissionBroker` function uses first element in the `brokersToDecommission` array. Why?